File: mobile/src/types/order.ts
/** 주문 Stage 2 API 계약 */

export type OrderListItem = {
  order_no: string
  order_dt: string
  custm_id: string
  customer: string
  status_cd: string
  status_nm: string
  total_qty: number
  total_amt: number
  pre_pay_amt: number
  line_count?: number
  rep_item_cd?: string
  rep_variety_cd?: string
  rep_variety_nm?: string
  rep_grade_cd?: string
  rep_grade_nm?: string
  rep_size_cd?: string
  rep_size_nm?: string
  rep_weight?: number
  delivery_tp_cd?: string
  delivery_tp_nm?: string
  delivery_tp_count?: number
  confirmed_shipped_qty?: number
  remaining_order_qty?: number
}

export type OrderListQuery = {
  from_date?: string
  to_date?: string
  status_cd?: string
  keyword?: string
  page?: number
  page_size?: number
}

export type OrderListPage = {
  items: OrderListItem[]
  total: number
  page: number
  page_size: number
}

export type OrderDelivery = {
  order_dlvry_id: string
  order_detail_id: string
  delivery_tp_cd: string
  qty: number
  planned_dt: string
  snd_name: string
  snd_tel: string
  snd_addr: string
  rcv_name: string
  rcv_tel: string
  rcv_addr: string
  dlvry_msg?: string
  delivery_tp_nm?: string
  /** 이 배송지로 이미 확정 출고된 수량 */
  confirmed_shipped_qty?: number
  /** 이 배송지의 남은 출고 대상 수량 */
  remaining_qty?: number
}

export type OrderLine = {
  order_detail_id: string
  item_cd: string
  variety_cd: string
  grade_cd: string
  size_cd: string
  weight: number
  qty: number
  unit_price: number
  item_amt: number
  harvest_year: number
  wh_cd: string
  dlvry_tp: string
  variety_nm?: string
  grade_nm?: string
  size_nm?: string
  dlvry_tp_nm?: string
  allocated_qty?: number
  unallocated_qty?: number
  reserved_unshipped_qty?: number
  confirmed_shipped_qty?: number
  remaining_order_qty?: number
  /** order_dlvry_id 없이 출고된 수량. >0이면 배송지 자동 seed 금지 */
  untracked_delivery_shipped_qty?: number
  deliveries: OrderDelivery[]
}

export type OrderDetail = OrderListItem & {
  mobile: string
  stock_status: string
  season_type_cd: string
  tot_order_amt: number
  tot_ship_fee: number
  tot_pay_amt: number
  rmk: string
  sales_no: string
  lines: OrderLine[]
}

export type OrderDeliveryPayload = {
  delivery_tp_cd: string
  qty: number
  planned_dt?: string | null
  snd_name?: string
  snd_tel?: string
  snd_addr?: string
  rcv_name?: string
  rcv_tel?: string
  rcv_addr?: string
  dlvry_msg?: string
}

export type OrderLinePayload = {
  variety_cd: string
  weight: number
  grade_cd: string
  size_cd: string
  qty: number
  unit_price: number
  harvest_year?: number | null
  warehouse_cd?: string | null
  item_cd?: string | null
  dlvry_tp?: string | null
  deliveries: OrderDeliveryPayload[]
}

export type OrderCreatePayload = {
  custm_id: string
  order_dt?: string | null
  season_type_cd?: string
  pre_pay_amt?: number
  tot_ship_fee?: number
  rmk?: string
  lines: OrderLinePayload[]
}

export type CustomerListItem = {
  custm_id: string
  custm_nm: string
  mobile: string
}

export type CustomerCreatePayload = {
  custm_nm: string
  mobile: string
  addr1?: string
  addr2?: string
  rmk?: string
}

File: server/schemas/order.py
# -*- coding: utf-8 -*-
"""주문 Stage 2 API 스키마."""

from __future__ import annotations

from pydantic import BaseModel, ConfigDict, Field


class OrderDeliveryIn(BaseModel):
    model_config = ConfigDict(extra="forbid")

    delivery_tp_cd: str = Field(..., min_length=1)
    qty: float = Field(..., gt=0)
    planned_dt: str | None = None
    snd_name: str = ""
    snd_tel: str = ""
    snd_addr: str = ""
    rcv_name: str = ""
    rcv_tel: str = ""
    rcv_addr: str = ""
    dlvry_msg: str = ""


class OrderLineIn(BaseModel):
    model_config = ConfigDict(extra="forbid")

    variety_cd: str = Field(..., min_length=8, max_length=8)
    weight: float = Field(..., ge=0)
    grade_cd: str = Field(..., min_length=1)
    size_cd: str = Field(..., min_length=1)
    qty: float = Field(..., gt=0)
    unit_price: float = Field(..., ge=0)
    harvest_year: int | None = None
    warehouse_cd: str | None = None
    item_cd: str | None = None
    dlvry_tp: str | None = None
    deliveries: list[OrderDeliveryIn]


class OrderCreateRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    custm_id: str = Field(..., min_length=1)
    order_dt: str | None = None
    season_type_cd: str = ""
    pre_pay_amt: float = 0
    pre_pay_method_cd: str | None = None
    tot_ship_fee: float = 0
    rmk: str = ""
    lines: list[OrderLineIn]


class OrderDeliveryOut(BaseModel):
    model_config = ConfigDict(extra="forbid")

    order_dlvry_id: str
    order_detail_id: str
    delivery_tp_cd: str
    qty: float
    planned_dt: str
    snd_name: str = ""
    snd_tel: str = ""
    snd_addr: str = ""
    rcv_name: str = ""
    rcv_tel: str = ""
    rcv_addr: str = ""
    dlvry_msg: str = ""
    delivery_tp_nm: str = ""
    confirmed_shipped_qty: float = 0
    remaining_qty: float = 0


class OrderLineOut(BaseModel):
    model_config = ConfigDict(extra="forbid")

    order_detail_id: str
    item_cd: str
    variety_cd: str
    grade_cd: str
    size_cd: str
    weight: float
    qty: float
    unit_price: float
    item_amt: float
    harvest_year: int
    wh_cd: str
    dlvry_tp: str
    variety_nm: str = ""
    grade_nm: str = ""
    size_nm: str = ""
    dlvry_tp_nm: str = ""
    allocated_qty: float = 0
    unallocated_qty: float = 0
    reserved_unshipped_qty: float = 0
    confirmed_shipped_qty: float = 0
    remaining_order_qty: float = 0
    untracked_delivery_shipped_qty: float = 0
    deliveries: list[OrderDeliveryOut]


class OrderListItem(BaseModel):
    model_config = ConfigDict(extra="forbid")

    order_no: str
    order_dt: str
    custm_id: str
    customer: str
    status_cd: str
    status_nm: str
    total_qty: float
    total_amt: float
    pre_pay_amt: float
    # compact list 보강 (상세 호환을 위해 default)
    line_count: int = 0
    rep_item_cd: str = ""
    rep_variety_cd: str = ""
    rep_variety_nm: str = ""
    rep_grade_cd: str = ""
    rep_grade_nm: str = ""
    rep_size_cd: str = ""
    rep_size_nm: str = ""
    rep_weight: float = 0
    delivery_tp_cd: str = ""
    delivery_tp_nm: str = ""
    delivery_tp_count: int = 0
    confirmed_shipped_qty: float = 0
    remaining_order_qty: float = 0


class OrderListPage(BaseModel):
    model_config = ConfigDict(extra="forbid")

    items: list[OrderListItem]
    total: int
    page: int
    page_size: int


class OrderDetail(OrderListItem):
    mobile: str = ""
    stock_status: str = "N"
    season_type_cd: str = ""
    tot_order_amt: float = 0
    tot_ship_fee: float = 0
    tot_pay_amt: float = 0
    pre_pay_method_cd: str | None = None
    rmk: str = ""
    sales_no: str = ""
    lines: list[OrderLineOut]


class CustomerCreateRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    custm_nm: str = Field(..., min_length=1)
    mobile: str = Field(..., min_length=1)
    addr1: str = ""
    addr2: str = ""
    rmk: str = ""


class CustomerListItem(BaseModel):
    model_config = ConfigDict(extra="forbid")

    custm_id: str
    custm_nm: str
    mobile: str = ""


class AllocationStockOut(BaseModel):
    model_config = ConfigDict(extra="forbid")

    alloc_id: str
    wh_cd: str
    item_cd: str
    variety_cd: str
    grade_cd: str
    size_cd: str
    weight: float
    harvest_year: int
    storage_dt: str
    allocated_qty: float
    shipped_qty: float = 0


class AllocationDetailOut(BaseModel):
    model_config = ConfigDict(extra="forbid")

    order_detail_id: str
    order_qty: float
    allocated_qty: float
    unallocated_qty: float
    reserved_unshipped_qty: float
    allocations: list[AllocationStockOut]


class AllocationSummaryOut(BaseModel):
    model_config = ConfigDict(extra="forbid")

    order_no: str
    farm_cd: str
    details: list[AllocationDetailOut]


class AllocationCreateRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    order_detail_id: str = Field(..., min_length=1)
    qty: float | None = Field(default=None, gt=0)
    auto: bool = False


class AllocationReleaseRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    order_detail_id: str = Field(..., min_length=1)
    qty: float = Field(..., gt=0)


class FruitStockItemOut(BaseModel):
    model_config = ConfigDict(extra="forbid")

    farm_cd: str
    wh_cd: str
    item_cd: str
    item_nm: str = ""
    variety_cd: str
    variety_nm: str = ""
    grade_cd: str
    grade_nm: str = ""
    size_cd: str
    size_nm: str = ""
    weight: float
    harvest_year: int
    storage_dt: str
    in_qty: float
    out_qty: float
    real_qty: float
    reserved_qty: float
    available_qty: float


class StockLogOut(BaseModel):
    """재고 이력 1건."""
    log_id: int
    farm_cd: str
    item_cd: str
    variety_cd: str
    variety_nm: str = ""
    harvest_year: int
    grade_cd: str
    grade_nm: str = ""
    size_cd: str
    size_nm: str = ""
    weight: float
    io_type: str
    io_type_nm: str = ""    # 사람이 이해하는 명칭 (생산입고, 원물사용, 주문배정 등)
    qty: float
    remark: str = ""
    reg_id: str = ""
    reg_dt: str = ""
